cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,619 @@
+
+data MonitorStateFileSet
+   = MonitorStateFileSet !(Map FilePath MonitorStateFile)
+                         ![FileGlobMonitorState]
+  deriving Show
+
+--instance Monoid MonitorStateFileSet where
+--  mempty = MonitorStateFileSet Map.empty []
+--  MonitorStateFileSet a b `mappend` MonitorStateFileSet x y =
+--    MonitorStateFileSet (a<>x) (b<>y)
+
+type Hash = Int
+type ModTime = UTCTime
+
+-- | The state necessary to determine whether monitored files have changed.
+--
+-- This covers all the cases of 'MonitorFilePath' except for globs which is
+-- covered separately by 'FileGlobMonitorState'.
+--
+data MonitorStateFile
+   = MonitorStateFile       !ModTime         -- ^ cached file mtime
+   | MonitorStateFileHashed !ModTime !Hash   -- ^ cached file mtime and content hash
+   | MonitorStateFileNonExistent
+
+     -- | These two are to deal with the situation where we've been asked
+     -- to monitor a file that's expected to exist, but when we come to
+     -- check it's status, it no longer exists.
+   | MonitorStateFileChanged
+   | MonitorStateFileHashChanged
+  deriving (Show, Generic)
+
+instance Binary MonitorStateFile
+
+-- | The state necessary to determine whether the files matched by a globbing
+-- match have changed.
+--
+data FileGlobMonitorState
+   = MonitorStateGlobDirs  !Glob !FilePathGlob
+                           !ModTime
+                           ![(FilePath, FileGlobMonitorState)] -- invariant: sorted
+
+   | MonitorStateGlobFiles !Glob
+                           !ModTime
+                           ![(FilePath, ModTime, Hash)] -- invariant: sorted
+  deriving (Show, Generic)
+
+instance Binary FileGlobMonitorState
+
+-- We can build a 'MonitorStateFileSet' from a set of 'MonitorFilePath' by
+-- inspecting the state of the file system, and we can go in the reverse
+-- direction by just forgetting the extra info.
+--
+reconstructMonitorFilePaths :: MonitorStateFileSet -> [MonitorFilePath]
+reconstructMonitorFilePaths (MonitorStateFileSet singlePaths globPaths) =
+    Map.foldrWithKey (\k x r -> getSinglePath k x : r)
+                     (map getGlobPath globPaths)
+                     singlePaths
+  where
+    getSinglePath filepath monitorState =
+      case monitorState of
+        MonitorStateFile{}          -> MonitorFile            filepath
+        MonitorStateFileHashed{}    -> MonitorFileHashed      filepath
+        MonitorStateFileNonExistent -> MonitorNonExistentFile filepath
+        MonitorStateFileChanged     -> MonitorFile            filepath
+        MonitorStateFileHashChanged -> MonitorFileHashed      filepath
+
+    getGlobPath (MonitorStateGlobDirs  glob globs _ _) =
+      MonitorFileGlob (GlobDir  glob globs)
+    getGlobPath (MonitorStateGlobFiles glob       _ _) =
+      MonitorFileGlob (GlobFile glob)
+
+------------------------------------------------------------------------------
+-- Checking the status of monitored files
+--
+
+data Changed b = Unchanged b | Changed (Maybe FilePath)
```

Comment:
I don't understand this type. When is Changed Nothing return? When the file went away? What does the file represent? The usage immediately below doesn't illuminate much either.